compiler/optimizer/abstractinterpreter/IDT.hpp

```diff
@@ -0,0 +1,128 @@
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef IDT_INCL
+#define IDT_INCL
+
+#include "compile/Compilation.hpp"
+#include "optimizer/CallInfo.hpp"
+#include "optimizer/abstractinterpreter/IDTNode.hpp"
+#include "env/Region.hpp"
+#include "env/VerboseLog.hpp"
+#include "il/ResolvedMethodSymbol.hpp"
+#include <queue>
+
+namespace TR {
+
+/**
+ * IDT stands for Inlining Dependency Tree
+ * It is a structure that holds all candidate methods to be inlined. 
+ * 
+ * The parent-child relationship in the IDT corresponds to the caller-callee relationship.
+ */
+class IDT
+   {
+   public:
+   IDT(TR::Region& region, TR_CallTarget*, TR::ResolvedMethodSymbol* symbol, uint32_t budget, TR::Compilation* comp);
+
+   TR::IDTNode* getRoot() { return _root; };
+
+   TR::Region& getRegion() { return _region; }
+
+   void addCost(uint32_t cost) { _totalCost += cost; }
+   uint32_t getTotalCost() { return _totalCost; }
+   
+   /**
+    * @brief Get the total number of nodes in this IDT.
+    *
+    * @return the total number of node
+    */
+   uint32_t getNumNodes() { return _maxIdx + 1; }
+   
+   /**
+    * @brief Get the next avaible IDTNode index.
+    *
+    * @return the next index
+    */
+   int32_t getNextGlobalIDTNodeIndex() { return _maxIdx; }
+
+   /**
+    * @brief Increase the next available IDTNode index by 1.
+    * This should only be called when successfully adding an IDTNode to the IDT
+    */
+   void increaseGlobalIDTNodeIndex()  { _maxIdx ++; }
+
+   /**
+    * @brief Get the IDTNode using index.
+    * Before using this method for accessing IDTNode, flattenIDT() must be called.
+    *
+    * @return the IDT node
+    */
+   TR::IDTNode *getNodeByGlobalIndex(int32_t index);
+   
+   /**
+    * @brief Flatten all the IDTNodes into a list.
+    */
+   void flattenIDT();
+
+   void print();
+
+   private:
+   TR::Compilation* comp() { return _comp; }
+
+   TR::Compilation *_comp;
+   TR::Region&  _region;
+   int32_t _maxIdx;
+   uint32_t _totalCost;
+   TR::IDTNode* _root;
+   TR::IDTNode** _indices;
+   };
+
+/**
+ * Accessing IDTNode by the priority of its cost.
+ */
+class IDTPreorderPriorityQueue
```

Comment:
You are correct.  This is not a pre-order.  In fact, the order in the queue is actually a reverse post-order since we want to prevent missing any parent before interpreting a node.  

This interleaving will not affect the algorithm.

Reply:
In that case, would you mind correcting the terminology in the name or at least in the comment? It would make the code clearer

If we're considering this not to be in preorder due to the interleaving I mentioned, then I don't think it's in reverse postorder either, by the way. It should be a topological sort though - in DAGs, topological sort is more general than reverse postorder

Reply:
The comment has been added to the PriorityQueue

Reply:
The comments of `IDTPriorityQueue` has been rewritten now. It should be clear enough now.

Actually, this part is **pre-ordered** and the interleaving is intended to do so.
I have checked code, this `IDTPriorityQueue` hasn't been called by phase 2 but by a part of code in phase 3, which implements a dynamic programming method to choosing which IDT node should be inlined based on their cumulative benefit and cumulative cost.
This DP method was implemented based on a patent of IBM, which is solving a nested knapsack problem for inlining, and here is the reference:
[https://patents.google.com/patent/US10055210B2/en](url)

This `IDTPriorityQueue` was called pre-order is because it accesses the IDT in pre-order to make sure all ancestors have been visited before the decedents. The interleaving is designed for the DP algorithm of the patent.

I have changed the comparator of the priority queue to first choosing the node that has a lower benefit, then choosing the one has lowest cost to break the tie since the patent mentioned:
>constraints on the order of node consideration, where the constraints serve to ensure every node is considered only after all of the node's ancestors are considered, and nodes are included in order of increasing cumulative benefit with lowest cumulative cost used to break ties to allow the generation of all intermediate solutions at a given cost budget to simplify backtracking in subsequent iterations of the algorithm

BTW, the reverse post-order was used in another part of code in OpenJ9 as a block iterator for CFG and it is a prebuilt method, so it shouldn't have any problem (here is a link for the reverse post-order algorithm, and basically it has same usage with the topological sorting, https://stackoverflow.com/questions/36131500/what-is-the-reverse-postorder).